File: packages/document-gateway/src/document-map.ts
import type { DocumentId, DocumentVersion } from './types'

export const DOCUMENT_MAP_CONTRACT_VERSION = 'document-map-v1' as const

export type DocumentMapBlockKind =
  'paragraph' | 'heading' | 'listItem' | 'table' | 'figure' | 'other'

export type DocumentMapFigureType = 'image' | 'chart' | 'other'

export interface DocumentMapLocator {
  readonly documentId: DocumentId
  readonly version: DocumentVersion
  readonly blockId: string
  readonly blockOrdinal: number
  readonly docxIndex?: number
  readonly sectionId?: string
}

export interface DocumentMapSection {
  readonly id: string
  readonly headingText: string
  readonly level: number
  readonly parentId?: string
  readonly locator: DocumentMapLocator
  readonly blockIds: readonly string[]
  readonly isDeleted: boolean
}

export interface DocumentMapBlock {
  readonly id: string
  readonly ordinal: number
  readonly kind: DocumentMapBlockKind
  readonly text: string
  readonly locator: DocumentMapLocator
  readonly sectionId?: string
  readonly headingLevel?: number
  readonly caption?: string
  readonly isDeleted: boolean
}

export interface DocumentMapTable {
  readonly id: string
  readonly locator: DocumentMapLocator
  readonly rowCount: number
  readonly columnCount: number
  readonly caption?: string
}

export interface DocumentMapFigure {
  readonly id: string
  readonly locator: DocumentMapLocator
  readonly figureType: DocumentMapFigureType
  readonly caption?: string
}

export interface DocumentMapCitation {
  readonly id: string
  readonly locator: DocumentMapLocator
  readonly text: string
  /** Unicode scalar/code-point offsets within the containing block text. */
  readonly start: number
  readonly end: number
  readonly format?: string
}

export interface DocumentMapReference {
  readonly id: string
  readonly locator: DocumentMapLocator
  readonly text: string
}

/** Provider-neutral structural snapshot of one active manuscript version. */
export interface DocumentMap {
  readonly contractVersion: typeof DOCUMENT_MAP_CONTRACT_VERSION
  readonly documentId: DocumentId
  readonly version: DocumentVersion
  readonly sections: readonly DocumentMapSection[]
  readonly blocks: readonly DocumentMapBlock[]
  readonly tables: readonly DocumentMapTable[]
  readonly figures: readonly DocumentMapFigure[]
  readonly citations: readonly DocumentMapCitation[]
  readonly references: readonly DocumentMapReference[]
}

export function isDocumentMapCurrent(
  map: DocumentMap,
  documentId: DocumentId,
  version: DocumentVersion,
): boolean {
  return map.documentId === documentId && map.version === version
}

export function isDocumentMapStale(
  map: DocumentMap,
  documentId: DocumentId,
  version: DocumentVersion,
): boolean {
  return !isDocumentMapCurrent(map, documentId, version)
}

File: 9profs-core-rs/crates/nineprofs-research/src/document_map.rs
use serde::{Deserialize, Serialize};

pub const DOCUMENT_MAP_CONTRACT_VERSION: &str = "document-map-v1";

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMap {
    pub contract_version: String,
    pub document_id: String,
    pub version: i64,
    pub sections: Vec<DocumentMapSection>,
    pub blocks: Vec<DocumentMapBlock>,
    pub tables: Vec<DocumentMapTable>,
    pub figures: Vec<DocumentMapFigure>,
    pub citations: Vec<DocumentMapCitation>,
    pub references: Vec<DocumentMapReference>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMapLocator {
    pub document_id: String,
    pub version: i64,
    pub block_id: String,
    pub block_ordinal: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub docx_index: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub section_id: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMapSection {
    pub id: String,
    pub heading_text: String,
    pub level: u8,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<String>,
    pub locator: DocumentMapLocator,
    pub block_ids: Vec<String>,
    pub is_deleted: bool,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMapBlock {
    pub id: String,
    pub ordinal: u32,
    pub kind: DocumentMapBlockKind,
    pub text: String,
    pub locator: DocumentMapLocator,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub section_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub heading_level: Option<u8>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub caption: Option<String>,
    pub is_deleted: bool,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum DocumentMapBlockKind {
    Paragraph,
    Heading,
    ListItem,
    Table,
    Figure,
    Other,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMapTable {
    pub id: String,
    pub locator: DocumentMapLocator,
    pub row_count: u32,
    pub column_count: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub caption: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMapFigure {
    pub id: String,
    pub locator: DocumentMapLocator,
    pub figure_type: DocumentMapFigureType,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub caption: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum DocumentMapFigureType {
    Image,
    Chart,
    Other,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMapCitation {
    pub id: String,
    pub locator: DocumentMapLocator,
    pub text: String,
    /// Unicode scalar/code-point offsets within the containing block text.
    pub start: u32,
    pub end: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub format: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentMapReference {
    pub id: String,
    pub locator: DocumentMapLocator,
    pub text: String,
}

pub fn is_document_map_current(map: &DocumentMap, document_id: &str, version: i64) -> bool {
    map.document_id == document_id && map.version == version
}

pub fn is_document_map_stale(map: &DocumentMap, document_id: &str, version: i64) -> bool {
    !is_document_map_current(map, document_id, version)
}
